examples/merge.c

```diff
@@ -0,0 +1,355 @@
+		const char *curr = argv[args.pos];
+
+		if (curr[0] != '-') {
+			opts_add_refish(opts, curr);
+		} else if (!strcmp(curr, "--no-commit")) {
+			opts->no_commit = 1;
+		} else if (match_str_arg(repo_path, &args, "--git-dir")) {
+		} else {
+			print_usage();
+		}
+	}
+}
+
+int resolve_refish(git_annotated_commit **commit, git_repository *repo, const char *refish)
+{
+	git_reference *ref;
+	int err = 0;
+	git_oid oid;
+
+	assert(commit != NULL);
+
+	err = git_reference_dwim(&ref, repo, refish);
+	if (err == GIT_OK) {
+		git_annotated_commit_from_ref(commit, repo, ref);
+		git_reference_free(ref);
+		return 0;
+	}
+
+	err = git_oid_fromstr(&oid, refish);
+	if (err == GIT_OK) {
+		err = git_annotated_commit_lookup(commit, repo, &oid);
+	}
+
+	return err;
+}
+
+int resolve_heads(git_repository *repo, merge_options *opts) {
+	git_annotated_commit **annotated = calloc(opts->heads_count, sizeof(git_annotated_commit *));
+	size_t annotated_count = 0, i;
+	int err = 0;
+
+	for (i = 0; i < opts->heads_count; i++) {
+		err = resolve_refish((git_annotated_commit **)&annotated[annotated_count++], repo, opts->heads[i]);
+		if (err != 0) {
+			fprintf(stderr, "failed to resolve refish %s (%d)\n", opts->heads[i], err);
+			annotated_count--;
+			continue;
+		}
+	}
+
+	if (annotated_count <= 0) {
+		fprintf(stderr, "unable to parse any refish\n");
+		free(annotated);
+		return -1;
+	}
+
+	opts->annotated = annotated;
+	opts->annotated_count = annotated_count;
+	return 0;
+}
+
+int perform_fastforward(git_repository *repo, const git_oid *target_oid, int is_unborn)
+{
+	git_checkout_options ff_checkout_options = GIT_CHECKOUT_OPTIONS_INIT;
+	git_reference *head_ref;
+	git_reference *new_head_ref;
+	int err = 0;
+
+
+	if (is_unborn) {
+		const char *symbolic_ref;
+		git_reference *master_ref;
+
+		/* master reference is unborn, lookup HEAD manually, */
+		err = git_reference_lookup(&head_ref, repo, "HEAD");
+		if (err != 0) {
+			fprintf(stderr, "failed to lookup HEAD ref\n");
+			return -1;
+		}
+
+		/* grab its target ref, */
+		symbolic_ref = git_reference_symbolic_target(head_ref);
+
+		/* force-create it on the targets' OID, */
+		err = git_reference_create(&master_ref, repo, symbolic_ref, target_oid, 1, NULL);
+		if (err != 0) {
+			fprintf(stderr, "failed to create master reference\n");
+			return -1;
+		}
+	} else {
+		/* The repo HEAD is pointing to an existing reference. */
+		err = git_repository_head(&head_ref, repo);
+		if (err != 0) {
+			fprintf(stderr, "failed to get HEAD reference\n");
+			return -1;
+		}
+
+		/* Move the HEAD reference to the new OID... */
+		err = git_reference_set_target(&new_head_ref, head_ref, target_oid, NULL);
+		if (err != 0) {
+			fprintf(stderr, "failed to move HEAD reference\n");
+			return -1;
+		}
+	}
+
+	/* force-checkout the result. */
+	err = git_checkout_head(repo, &ff_checkout_options);
+	if (err != 0) {
+		fprintf(stderr, "failed to checkout HEAD reference\n");
+		return -1;
+	}
+	return 0;
+}
```

Comment:
As long as we don't pretend to allow fast-forward merges (e.g. accepting the "--ff-only" flag, which I bet isn't part of this PR anyway) I don't mind it. The examples are probably not used by anybody anyway and only aim to give examples of how to use our API. Handling all the intricacies of upstream git tools does not work towards that goal and can even overcomplicate examples, working straight against the desire to demonstrate our API.